### DISCUSSION TITLE
Disable keyboard with a bind

### DISCUSSION THREAD

**Discussion by noahfraiture** [hyprwm/Hyprland#4283] [Q&A] [answered]
Hey, I search but didn't find a way to disable the whole keyboard with a bind. I found this [https://wiki.hyprland.org/Configuring/Uncommon-tips--tricks/](url) that allow us to disable binds, but not the whole keyboard.
This is a laptop and I would like to have a bind to disable it do put another keyboard on it or other things like my phone when I don't need the keyboard

**Comment by muellerbernd** [comment]
[This](https://www.reddit.com/r/hyprland/comments/11kr8bl/hotkey_disable_touchpad/) worked pretty well, but since #4656 was merged this script does not work anymore. Maybe someone can help to get the script working again.

**Reply by vaxerski** [reply]
if you have `hyprlang-git` installed you can use a variable like `$MYKEYBOARDON` and dynamically set it with `hyprctl keyword`

**Reply by muellerbernd** [reply]
I was using  [this script](https://github.com/muellerbernd/dotfiles/blob/main/hypr-common/.config/hypr/scripts/toggleKeyboard.sh) but now that does not work anymore. 
@vaxerski can you give an minimal example?

**Reply by vaxerski** [reply]
```
$AMONGUS_ENABLED = true
device {
    name = amongus
    enabled = $AMONGUS_ENABLED
}
```

command:
```sh
hyprctl keyword '$AMONGUS_ENABLED' "false"
```

remember to use `'` (single quotes) for the variable to tell your shell not to interpret it.

**Reply by muellerbernd** [reply]
I am using hyprland (via flake) on NixOS. For hyprlang I use the provided overlay (version 0.4.1). When I hardcode the variable `$AMONGUS_ENABLED` to `true` or `false` in the config and do a reload then the device (in my case the laptop keyboard) is correctly enabled or disabled. But when I try to set it dynamically via `hyprctl keyword '$AMONGUS_ENABLED' "false"` it does not change the state of the device but `hyprctl` returns `ok`. Can I access the current value of `$AMONGUS_ENABLED` via `hyprctl`?

**Reply by muellerbernd** [reply]
@vaxerski I tested setting the variable dynamically via `hyprctl` but it does not work. See my previous post.
The old way of doing this was `hyprctl keyword device:at-translated-set-2-keyboard:enabled false`.

**Reply by vaxerski** [reply]
you need hyprlang>=0.4.1 for it to work.
Also make sure to pass the `-r` opt to hyprctl

**Reply by muellerbernd** [reply]
@vaxerski  thanks for the hint with `-r` now it works. Is it possible to get the current value of the variable `e.g. $AMONGUS_ENABLED` via `hyprctl`?

**Reply by vaxerski** [reply]
nope

**Reply by muellerbernd** [reply]
I am using now this part in my config:
```
$LAPTOP_KB_ENABLED = true
device {
  name = at-translated-set-2-keyboard
  enabled = $LAPTOP_KB_ENABLED
}
```
@vaxerski is it possible to read the enabled option for devices via `hyprctl`?
The old way was `hyprctl getoption device:at-translated-set-2-keyboard:enabled`

**Reply by vaxerski** [reply]
nope, but maybe hyprctl devices lists it

**Comment by muellerbernd** [comment, marked as answer]
Here my current script for laptop keyboard toggle (inspired by [this reddit post](https://www.reddit.com/r/hyprland/comments/11kr8bl/hotkey_disable_touchpad/)):
```sh
#!/usr/bin/env bash

export STATUS_FILE="$XDG_RUNTIME_DIR/keyboard.status"

enable_keyboard() {
    printf "true" >"$STATUS_FILE"
    notify-send -u normal "Enabling Keyboard"
    hyprctl keyword '$LAPTOP_KB_ENABLED' "true" -r
}

disable_keyboard() {
    printf "false" >"$STATUS_FILE"
    notify-send -u normal "Disabling Keyboard"
    hyprctl keyword '$LAPTOP_KB_ENABLED' "false" -r
}

if ! [ -f "$STATUS_FILE" ]; then
  enable_keyboard
else
  if [ $(cat "$STATUS_FILE") = "true" ]; then
    disable_keyboard
  elif [ $(cat "$STATUS_FILE") = "false" ]; then
    enable_keyboard
  fi
fi
```
add the following to your hyprland config (the device name maybe depends on your machine):
```
$LAPTOP_KB_ENABLED = true
device {
  name = at-translated-set-2-keyboard
  enabled = $LAPTOP_KB_ENABLED
}
```

**Reply by newbiedeveloper968** [reply]
Sir, where do I add this code of yours?

"
#!/usr/bin/env bash

export STATUS_FILE="$XDG_RUNTIME_DIR/keyboard.status"

enable_keyboard() {
    printf "true" >"$STATUS_FILE"
    notify-send -u normal "Enabling Keyboard"
    hyprctl keyword '$LAPTOP_KB_ENABLED' "true" -r
}

disable_keyboard() {
    printf "false" >"$STATUS_FILE"
    notify-send -u normal "Disabling Keyboard"
    hyprctl keyword '$LAPTOP_KB_ENABLED' "false" -r
}

if ! [ -f "$STATUS_FILE" ]; then
  enable_keyboard
else
  if [ $(cat "$STATUS_FILE") = "true" ]; then
    disable_keyboard
  elif [ $(cat "$STATUS_FILE") = "false" ]; then
    enable_keyboard
  fi
fi
"

**Reply by muellerbernd** [reply]
You place it in a bash script and then you can call this via a keybinding.

**Comment by al23212321** [comment]
I am currently using 0.35.0-2, and I can disable the touchpad via 
`hyprctl keyword device:synps/2-synaptics-touchpad:enabled false`

where synps/2-synaptics-touchpad is the name of my device using hyprctl devices

**Reply by noahfraiture** [reply]
I get the error ```special category's first value must be the key. Key for <device> is <name>``` with my keyboard

**Reply by vaxerski** [reply]
because that won't work on 0.36+

**Reply by jenil1122** [reply]
hi i used the same command for my touchpad and the output was : 
```
❯ hyprctl keyword etps/2-elantech:enabled false
config option <etps/2-elantech:enabled> does not exist.
```

**Comment by Morbiuzx** [comment]
> hi i used the same command for my touchpad and the output was :
> 
> ```
> ❯ hyprctl keyword etps/2-elantech:enabled false
> config option <etps/2-elantech:enabled> does not exist.
> ```
I'm getting the exact same error, config option does not exist. Using hyprland-git 0.38.0.r77.a06272ae-1

**Comment by pjrulez** [comment]
Here's an alternative approach in ML4W Dotfiles context:

So, after trying all the wrong ways of doing it (xorg-xinput, evtest, ...), and almost breaking the whole configuration I finally got it.

Pasting here for reference if someone else wants to place your mechanical keyboard on top of your laptop without worry.

I'm using ML4W Dotfiles so my solution is within that context. Here are the necessary changes:

First, run '`hyprctl devices`', and find your builtin keyboard name. It will possibly be named something like "_at-translated-set-2-keyboard_"

Hyprland docs page that touches this topic: https://wiki.hyprland.org/Configuring/Keywords/#per-device-input-configs

Save this script somewhere in the waybar config folders like "_dotfiles/.config/waybar/toggle-laptop-keyboard.sh_":

```
#!/bin/bash
#~/.config/waybar/toggle-laptop-keyboard.sh &
#  _____                 _             
# |_   _|__   __ _  __ _| | ___  
#   | |/ _ \ / _` |/ _` | |/ _ \ 
#   | | (_) | (_| | (_| | |  __/    
#   |_|\___/ \__, |\__, |_|\___|    
#            |___/ |___/                          
#

CACHE_FILE="/home/USER/.cache/laptop-keyboard-disabled"

if [ -f "$CACHE_FILE" ] ;then
    rm "$CACHE_FILE"
    sh -c "hyprctl keyword 'device[at-translated-set-2-keyboard]:enabled' 1"
else
    touch "$CACHE_FILE"
    sh -c "hyprctl keyword 'device[at-translated-set-2-keyboard]:enabled' 0"
fi
```

This will do the job when executed.
The rest of my post is about how to add a toggle button to waybar.

You need to open 3 more files:

    Your waybar modules definition configuration file:
    _/home/USER/dotfiles/.config/waybar/modules.json_

Add this custom module definition to your modules.json:

 ```
 // Toggle Laptop Keyboard
  "custom/toggle-laptop-keyboard": {
    "format": "",
    "on-click": "~/.config/waybar/toggle-laptop-keyboard.sh",
    "tooltip-format": "Toggle Laptop Keyboard"
  },
```

And then add your new module to an existing group like in the example within modules.json like for example:

```
  // Group Links
  "group/links": {
    "orientation": "horizontal",
    "modules": [
      "custom/chatgpt",
      "custom/empty",
      "custom/toggle-laptop-keyboard" // like this
    ]
  },
```

OR within 2. Your chosen waybar theme config file:
_/home/USER/.ml4w-hyprland/backup/dotfiles/.config/waybar/themes/ml4w-modern/config_

for example like this:

```
    // Modules Left
    "modules-left": [
        "custom/appmenu",
        "hyprland/workspaces",
        "group/links",
        // "custom/mpd",
        // "custom/mpris",
        "custom/toggle-laptop-keyboard" // or like this
        //"wlr/taskbar",
        "group/quicklinks",
        "custom/empty"
    ],
```

3. Your chosen waybar theme style.css file:
_/home/USER/dotfiles/.config/waybar/themes/ml4w-modern/style.css_

Add `#custom-toggle-laptop-keyboard`, under '_Custom Quicklinks_' section to make sure your toggle button is applied the same styling as the rest of your waybar buttons.

Now you need to open the ML4W Settings App and turn off and turn back on Waybar. Maybe you'll have to run hyprctl reload as well but probably it will work right away!

![image](https://github.com/user-attachments/assets/7de8f742-789a-427f-b24e-c9e00b53c594)
_Font Awesome Keyboard Icon Enable/Disable Input Device Toggle Button in Hyprland /w ML4W Dotfiles_

The same idea can be applied to enable/disable other input devices and you can follow the same approach to implement an easy to use solution in your customized configuration

https://gist.github.com/pjrulez/a0a1e5c06b3579cbf637101a0cc07f8d

**Comment by ChristofferNissen** [comment]
For my workflow this is all i needed:

```hyprland
$LAPTOP_KB_ENABLED = true
device {
    name = at-translated-set-2-keyboard
    enabled = $LAPTOP_KB_ENABLED
}
```

Then if i am using my external keyboard, i just execute:

`hyprctl keyword '$LAPTOP_KB_ENABLED' "false" -r`
